ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,464 @@
+| AnnTargetProvisioner      | provionser.appstudio.redhat.com/dt-provisioner | A name of a provisioner | Binding controller | DTC          | Indicates that a DT should be dynamically provisioned for the DTC by the provisioner whose name appear in the value                                      |
+| AnnBindCompleted          | dt.appstudio.redhat.com/bind-complete          | "yes"                   | Binding controller | DTC          | Indicates that the binding controller completed the binding process                                                                                      |
+| AnnBoundByController      | dt.appstudio.redhat.com/bound-by-controller    | "yes"                   | Binding controller | DTC          | Indicates that the binding controller bind the DTC to a DT. In practice it means that the controller set the DTC.spec.VolumeName to the value of DT.Name |
+| AnnDynamicallyProvisioned | provionser.appstudio.redhat.com/provisioned-by | A name of a provisioner | Provisioner        | DT           | Indicates that the provisioner whose name appears in the value provisioned the DT.                                                                       |
+
+### Finalizer details
+
+| Const Name in code | Key                                       | Applied by  | Applied on | Purpose                                                                       |
+|--------------------|-------------------------------------------|-------------|------------|-------------------------------------------------------------------------------|
+| finalizerDT        | provionser.appstudio.redhat.com/finalizer | Provisioner | DT         | Delays the deletion of a DT so the provisioner can free up external resources |
+
+### Use Case Descriptions
+
+- **During onboarding** - when a user requests a new appstudio tier namespace, the tier template includes two [Environments], and two [DTCs]. The [Environments] reference the [DTCs]. The [DTCs] bear a request for the `devsandbox` [DTCLS]. The devsandbox provisioner responds to that request and generates a `SpaceRequest`, ultimately resulting in a new namespace for each environment. The spacerequest controller provisions a serviceaccount in the new namespace and places a kubeconfig including the token for this serviceaccount in the origin namespace. The `SpaceRequest` in the origin namespace is updated to reference that `Secret` and is marked ready by the `SpaceRequest` controller. The devsandbox deployment target provisioner controller sees that and creates the devsandbox [DT] referencing the `Secret` and marks the devsandbox [DT] as ready. The deployment target binder sees that, and attaches the new [DTs] to the [DTCs]. The environment controller sees this and creates a [GitOpsDeploymentManagedEnvironment] that references the `Secret` found from traversing from the [Environment] to the [DTC] to the [DT].
```

Comment:
Instead of introducing a devsandbox provisioner and the SpaceRequest API/Controller : 
Could we implement a "free-tier-osd-provisioner" controller which takes a pool of OSDs and gets us namespaces ? The same could be used against Sandbox OSD clusters too, what do you think ?

Reply:
Yes, although the provisioner needs to be coded to manage specific kinds of external resources. The devsandbox provisioner described here knows specifically how to create and manage `SpaceRequests`. If we want to get namespaces from another pool of OSD clusters that are _not_ managed by codeready toolchain operators, then we'll need a provisioner that knows how to do that _without_ `SpaceRequests`. The proposal gives us freedom to implement one or both, exposing each as different `DeploymentTargetClasses`.

FWIW, we already have devsandbox member clusters available and assuming that the SpaceRequest API really will be implemented - I think we should build against it for the first provisioner implementation.

Reply:
> Could we implement a "free-tier-osd-provisioner" controller which takes a pool of OSDs and gets us namespaces 

This is basically what Dev Sandbox does in nutshell :)
So, you are asking if instead of using Dev Sandbox we could re-implement dev sandbox in the "free-tier-osd-provisioner". Of course we could. But why? :)

Reply:
> we could re-implement dev sandbox in the "free-tier-osd-provisioner". Of course we could. But why? :) 

So customers could create private "devsandboxes" the same way WE do as part of a BYOC cluster pool they want to control ? Otherwise our devsandbox is provisioned/configured in a special way and differently than the perhaps customers would try to create their own free-tier thing.  Why do we need two ways to do it ?

Reply:
> if we want to get namespaces from another pool of OSD clusters that are not managed by codeready toolchain operators, then we'll need a provisioner that knows how to do that without SpaceRequests.

Yes, this ^

My reframed question is, should we implement a provisioner that is generic enough to be used for provisioning namespaces from devsandbox OSDs / non-devsandbox OSDs ( or even non-openshift clusters ). How significant would that effort be ? We'll need to do that anyway sooner than later.

Reply:
Sandbox already provide a lot of useful functionality which such a provisioner would need to re-implement. SpaceRequest API is our next step towards making Sandbox is even more generic and flexible so it can be used as "Namespace as a Service" in more use cases, such as the namespace provisioner to create/manage namespaces in different types of clusters (control plane clusters, workload clusters, etc).

It's not only about provisioning a namespace. We need a service which would provide and *manage* namespaces in a fleet of clusters. It's not that trivial.